### PR TITLE
Fixed #28126 -- Added GiST Index to contrib.postgres.

### PR DESCRIPTION
**Pull request by mjtamlyn**
We can already create them on spatial fields, but they are generally useful for other fields such as hstore.

https://www.postgresql.org/docs/9.5/static/gist-examples.html
https://code.djangoproject.com/ticket/28126

### CODE REVIEW

File: docs/ref/contrib/postgres/indexes.txt [hunk 1]
```diff
@@ -34,3 +34,11 @@ available from the ``django.contrib.postgres.indexes`` module.
     PostgreSQL. You can install it using the
     :class:`~django.contrib.postgres.operations.BtreeGinExtension` migration
     operation.
+
+``GinIndex``
```
Comment (Stranger6667) [inline docs/ref/contrib/postgres/indexes.txt]: It should be ``GistIndex``, I guess

File: docs/ref/contrib/postgres/indexes.txt [hunk 2]
```diff
@@ -34,3 +34,11 @@ available from the ``django.contrib.postgres.indexes`` module.
...
+============
+
+.. class:: GinIndex()
```
Comment (Stranger6667) [inline docs/ref/contrib/postgres/indexes.txt]: As well as here

File: django/contrib/postgres/indexes.py [hunk 3]
```diff
@@ -45,3 +45,10 @@ class GinIndex(Index):
 
     def create_sql(self, model, schema_editor):
         return super().create_sql(model, schema_editor, using=' USING gin')
+
+
+class GistIndex(Index):
```
Comment (felixxm) [inline django/contrib/postgres/indexes.py]: Please add `GistIndex` to `__all__`.

**Comment by mjtamlyn** [comment]
Updated :)

File: docs/ref/contrib/postgres/indexes.txt [hunk 4]
```diff
@@ -34,3 +34,17 @@ available from the ``django.contrib.postgres.indexes`` module.
...
+
+.. versionadded:: 2.0
+
+.. class:: GistIndex()
+
+    Creates a `GiST index
+    <https://www.postgresql.org/docs/current/static/gist.html>`_. These indexes
+    are also created on spatial fields with the
+    :attr:`~django.contrib.gis.db.models.BaseSpatialField.spatial_index` when
+    using postgis, but are also useful on some other types such as
```
Comment (felixxm) [inline docs/ref/contrib/postgres/indexes.txt]: `postgis` -> `PostGIS`

**Comment by atombrella** [comment]
The indexes in PostgreSQL take some optional parameters. For GIST, this is `fillfactor` and `buffering`. Would it be an idea to include them as parameters?

https://www.postgresql.org/docs/current/static/sql-createindex.html

**Comment by timgraham** [comment]
A test in `SchemaTests` is also needed.

**Comment by timgraham** [comment]
Updated in https://github.com/django/django/pull/8694.

**Pull request closed without merging**